### ISSUE TITLE
"sendpin" command hangs when connected to a Trezor Model One

### ISSUE THREAD

**Issue by nadahalli** [bitcoin-core/HWI#636]
Commands to reproduce the problem.

```
$./hwi enumerate
[{"type": "trezor", "path": "webusb:001:3", "label": "My Trezor", "model": "trezor_1", "needs_pin_sent": true, "needs_passphrase_sent": true, "error": "Could not open client or get fingerprint information: Trezor is locked. Unlock by using 'promptpin' and then 'sendpin'.", "code": -12}]
```
```
$./hwi --device-type trezor promptpin
Use 'sendpin' to provide the number positions for the PIN as displayed on your device's screen
Use the numeric keypad to describe number positions. The layout is:
    7 8 9
    4 5 6
    1 2 3
{"success": true}
```
When this command runs, the device shows the keypad in a randomzied order, for example:

2 3 5
6 7 1
9 4 8

Say my pin is "1111", I believe I have to send the pin "9999" (real pin mapped to the randomized keypad).

`$./hwi --device-type trezor sendpin 9999`

This command hangs in the TrezorLib interruptRead command here: https://github.com/bitcoin-core/HWI/blob/d6a8cdef3b904a8a7ea5e316a83cb4037046202d/hwilib/devices/trezorlib/transport/webusb.py#L76

Any ideas what I am doing wrong with respect to the Trezor state machine protocol? Is there some other order in which promptpin and sendpin have to be used?

**Comment by prusnak** [comment]
> Say my pin is "1111", I believe I have to send the pin "9999"

No. To enter the PIN 1111 you have to enter 6666. (It's the other way around).

**Comment by prusnak** [comment]
Does it help if you reboot and try again? (Do not start Trezor Suite before!)

**Comment by nadahalli** [comment]
I can start again at "enumerate" if I disconnect and reconnect the hardware wallet. Promptpin works after that, but sendpin fails again. Is that what you mean?

**Comment by prusnak** [comment]
> Is that what you mean?

No, reboot is as in reboot your computer.

**Comment by prusnak** [comment]
Also, do you have the udev files installed? https://wiki.trezor.io/Udev_rules

**Comment by nadahalli** [comment]
Reboot of the computer doesn't help.
Yes, I have installed Trezor's udev rules. That doesn't seem to help either. 

Without "priming" the Trezor with the promptpin command, just sending a dummy sendpin command doesn't hang, but returns a success: false message.

```
$/hwi --device-type trezor sendpin 1111
{"success": false}
```
In other information, I have the following versions of the software.

```
$./hwi --version
hwi 2.1.1
$uname -rv
5.13.0-051300-generic #202106272333 SMP Sun Jun 27 23:36:43 UTC 2021
```

Trezor firmware version 1.11.2

**Comment by prusnak** [comment]
Any ideas @matejcik ?

**Comment by Whiskey992** [comment]
You can't hijack user account bitcoin will not allow it will lock everytime

**Comment by tracy434** [comment]
@

> Reboot of the computer doesn't help. Yes, I have installed Trezor's udev rules. That doesn't seem to help either.
> 
> Without "priming" the Trezor with the promptpin command, just sending a dummy sendpin command doesn't hang, but returns a success: false message.
> 
> ```
> $./hwi --device-type trezor sendpin 1111
> {"success": false}
> ```
> 
> In other information, I have the following versions of the software.
> 
> ```
> $./hwi --version
> hwi 2.1.1
> $uname -rv
> 5.13.0-051300-generic #202106272333 SMP Sun Jun 27 23:36:43 UTC 2021
> ```
> 
> Trezor firmware version 1.11.2

hello @nadahalli I will suggest you talk with a Trezor Developer concerning this issue,on their Trezor live chat Developer's support page. [click here](https://trezorforum.live/) Using your TICKET ID: #619589
to engage in a conversation with a representative from Trezor support and have the transaction looked into so as to resolve it right away.

**Comment by prusnak** [comment]
@achow101 can you please block the above user and remove their post? this is a phishing attempt.

**Comment by matejcik** [comment]
I can reproduce the problem locally.
It seems that when using `-t trezor` but not `-d <path>`, hwi tries to enumerate devices.
This tries to query the Trezor for its features.
That cannot be done while Trezor is waiting for PIN, so Trezor ignores the message, so you get stuck at reading from device.

**Comment by nadahalli** [comment]
Passing the device path works, as it prevents the find_device call.

```
$./hwi --device-type trezor promptpin
Use 'sendpin' to provide the number positions for the PIN as displayed on your device's screen
Use the numeric keypad to describe number positions. The layout is:
    7 8 9
    4 5 6
    1 2 3
{"success": true}
$./hwi --device-path webusb:001:3 --device-type trezor sendpin 5555
{"success": true}
```
The offending call is here: https://github.com/bitcoin-core/HWI/blob/e731395bde13362950e9f13e01689c475545e4dc/hwilib/_cli.py#L280, where without --device-path, HWI does a find_device. I guess it cannot be really fixed because if the path is not known, you have to find the device anyway. For now, I will just use both and make it work.

**Comment by harding** [comment]
I think this issue should be re-opened.  I just hit this today using a Trezor One for the first time with HWI.  As described in the OP, the following command is successful:

    hwi -t trezor promptpin

But this command will indefinitely hang:

    hwi -t trezor sendpin 1234567890

The only way to `sendpin` is by using both device and type:

    hwi -t trezor -d webusb:001:1 sendpin 1234567890

I think two changes should be made to avoid user confusion:

1. `sendpin` for Trezor devices should fail if `-d` has not been passed.  That's preferable to an indefinite hang and it's an opportunity to tell the user that they need to pass `-d`.
2. `promptpin` for Trezor devices should mention in its message that `-d` needs to passed for the subsequent `sendpin`.
